actionview/test/unit/cache_expiry_test.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+require "abstract_unit"
+
+class ActionViewCacheExpiryTest < ActiveSupport::TestCase
+  def test_execution_lock_can_be_acquired_and_released
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    # It can acquire read lock:
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+    execution_lock.release_read_lock
+    assert_equal 0, execution_lock.read_count
+
+    # It can acquire write lock:
+    execution_lock.with_write_lock do
+      assert_equal 1, execution_lock.write_count
+    end
+    assert_equal 0, execution_lock.write_count
+  end
+
+  def test_execution_lock_multiple_read_locks
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.acquire_read_lock
+    assert_equal 2, execution_lock.read_count
+
+    execution_lock.release_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.release_read_lock
+    assert_equal 0, execution_lock.read_count
+  end
+
+  def test_execution_lock_read_lock_upgraded_to_write_lock
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.with_write_lock do
+      assert_equal 1, execution_lock.write_count
+      assert_equal 1, execution_lock.read_count
+    end
+
+    assert_equal 1, execution_lock.read_count
+    execution_lock.release_read_lock
+
+    assert_equal 0, execution_lock.read_count
+  end
+
+  def test_execution_lock_write_lock_priority
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+    sequence = Thread::Queue.new
+
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    writer = Thread.new do
+      execution_lock.acquire_read_lock
+
+      execution_lock.with_write_lock do
+        sequence << :write_lock_acquired
+        assert_equal 1, execution_lock.write_count
+        assert_equal 1, execution_lock.read_count
+      end
+
+      execution_lock.release_read_lock
+    end
+
+    reader = Thread.new do
+      execution_lock.acquire_read_lock
+      sequence << :read_lock_acquired
+      assert_equal 1, execution_lock.read_count
+      execution_lock.release_read_lock
+      assert_equal 0, execution_lock.read_count
+    end
+
+    # Wait for both the threads to be waiting for the lock:
+    Thread.pass until reader.status == "sleep" && writer.status == "sleep"
```

Comment:
Why would the `reader` ever wait/sleep? It can acquire the read lock straight away and the same for the rest of its body, no?

Reply:
Ah, you are right, I should ensure the write lock has acquired first.